### PR TITLE
Allow configuration of indentation for YAML and JSON stores

### LINKED ISSUE #900: Configurable indentation for yaml output in .sops.yaml
It would be great if we could configure our indentation for yaml in the .sops.yaml configuration file. Right now it's just hardcoded to set the indent to 4 all the time.

https://github.com/mozilla/sops/blob/66043e71a81787d6513bc2e5505a29aac67dc6f1/stores/yaml/store.go#L322

Yaml spec does not dictate 4 spaces.
https://yaml.org/spec/1.2/spec.html#id2777534
```
In YAML block styles, structure is determined by indentation. In general, indentation is defined as a zero or more space characters at the start of a line.
```

### LINKED ISSUE #1028: To let `sops` emit YAML parser compatible .json files?
When `sops` is used to encrypt a .yaml file, the output is a .yaml file. When `sops` is used to encrypt any other file, the output is a `.json`file.

My colleagues were writing a check in Python to validate that all files that by a naming convention were supposed to be encrypted, was actually encrypted properly. To do that, we looked to parse them and look for a `sops` key as a basic check. Since all `sops` encrypted files are YAML or JSON, we figured we could use a YAML parser and it would work. But, it didn't. We had to use a YAML parser for the .yaml files and a JSON parser for the .json files.

It turns out that while almost any JSON is valid YAML, [JSON with tabs isn't valid YAML](https://yaml.org/faq.html). The `sops` generated .json files included tabs instead of spaces!

## Questions

__The main question__

Should `sops` keep emitting tabs in its emitted `.json` files, or could it switch to emitting spaces, and by doing so emit .json files that are both JSON and YAML parser compatible?

## Related code

From code inspection, this seemed relevant to understand how files are written.

https://github.com/mozilla/sops/blob/66043e71a81787d6513bc2e5505a29aac67dc6f1/stores/json/store.go#L281-L300

## Related issues

- https://github.com/mozilla/sops/issues/434

### LINKED ISSUE #1048: Allow for specifying the indent character to be used and how many (tabs OR spaces!)
Hello,

We are using sops as part of a pipeline that originally generates json pretty-formatted files that are space-indented. Unfortunately when we then use sops, [it decides](https://github.com/mozilla/sops/blob/66043e71a81787d6513bc2e5505a29aac67dc6f1/stores/json/store.go#L220) to indent using tabs, which can cause our diffs to be large and noisy.

We'd like sops to allow defining an indent character of choice and how many of those characters, similar to many other frameworks that work with JSON.

### PR DESCRIPTION
**Pull request by Ph0tonic**
This PR aims to finish the work started in #930, thanks to @jamesgoodhouse.

I added the use of the Indent param in the store as well as a new `--indent` option in the cli. 
Let me know what you think about it.

Fixes #900, fixes #1048 and fixes #1028

### CODE REVIEW

File: cmd/sops/main.go
Comment (devstein) [inline cmd/sops/main.go:1084]: Merge issue

Comment (Ph0tonic) [reply cmd/sops/main.go]: Ho thanks, I forgot to save my file before committing. Fixed !

**Comment by hiddeco** [comment]
In favor of adding support for this but not before 3.9.0.

File: cmd/sops/main.go
Comment (felixfontein) [inline cmd/sops/main.go:705]:
```suggestion
			Usage: "the number of space to indent YAML encoded file for encryption",
```

File: cmd/sops/main.go
Comment (felixfontein) [inline cmd/sops/main.go:1065]: It looks to me that lines 1058 to 1065 could be moved to a function, since they appear twice in this file (here and in `outputStore()`).

File: cmd/sops/common/common.go
Comment (felixfontein) [inline cmd/sops/common/common.go:35]: What is this used for? Maybe I missed something, but this never seems to be called, and all implementations don't do anything, not even the one for the YAML store.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've some first comments after glancing over the implementation.

File: cmd/sops/main.go
Comment (Ph0tonic) [reply cmd/sops/main.go:1065]: Changed

File: cmd/sops/common/common.go
Comment (Ph0tonic) [reply cmd/sops/common/common.go:35]: You are right, it's useless, good catch. It's a remainder of PR #930 from which I started. Maybe he had another idea in mind.

**Comment by felixfontein** [comment]
For some reason this doesn't compile in CI anymore. The only error printed is `Error: `, which is not exactly helpful. I'm not sure whether this short error output is caused by GHA or by Go itself. In another PR, #1274, the compile steps worked, but the functional tests fail, also with `Error: `, so this *might* be related to GHA. Did you try to compile this locally?

**Comment by felixfontein** [comment]
Meh, actually the problem was on my side - apparently GHA changed some code and the error reports are now loaded from another domain name, which I was blocking by default.

**Comment by Ph0tonic** [comment]
> Meh, actually the problem was on my side - apparently GHA changed some code and the error reports are now loaded from another domain name, which I was blocking by default.

I found an issue in `main.go` which I fixed. 👍

**Comment by felixfontein** [comment]
How about using tabs if no indent is specified, and the amount of spaces if `--indent` is specified (or the JSON indent value is configured in the config)?

It might also make sense to allow `0` as a valid value for JSON. (IIRC parts of the code assume that not specified == 0.)

**Comment by Ph0tonic** [comment]
> How about using tabs if no indent is specified, and the amount of spaces if `--indent` is specified (or the JSON indent value is configured in the config)?
> 
> It might also make sense to allow `0` as a valid value for JSON. (IIRC parts of the code assume that not specified == 0.)

Yes, we could do that, I think it would make sense.
I just see one technical problem if we want to allow 0, I do not think it's possible to differentiate whether zero was provided during the unmarshaling of `.sops.yaml` or if it was empty and got the default 0 value.

**Comment by Ph0tonic** [comment]
I applied what we discussed. However, while looking at other issues, #1028 seems to propose having default json indentation with spaces instead of tab to be compatible with yaml parser. I have no idea whether it is a good idea to change the default indentation of json to 4 spaces instead of one tab.
What do you think ?

**Comment by felixfontein** [comment]
> However, while looking at other issues, #1028 seems to propose having default json indentation with spaces instead of tab to be compatible with yaml parser. I have no idea whether it is a good idea to change the default indentation of json to 4 spaces instead of one tab.

I tend a bit to keep the tabs. No idea what the other maintainers think though :)

In any case, as long as it is possible to explicitly configure indentation by tabs (for example by setting indentation to a negative number) I'm ok with the default changing, assuming there is a majority for changing it.

> I just see one technical problem if we want to allow 0, I do not think it's possible to differentiate whether zero was provided during the unmarshaling of `.sops.yaml` or if it was empty and got the default 0 value.

I would expect it is possible to provide a default value. I'm not that familiar with Go's serialization/deserialization mechanism though...

**Comment by felixfontein** [comment]
I think you can simply assign default values to `conf` in https://github.com/getsops/sops/blob/d46b70272adf866fd25ff3f4356aa6c692acfaf6/config/config.go#L252-L253 before loading, like
```
	conf := &configFile{}
        conf.Stores.JSON.Indent = -1
	err = conf.load(confBytes)
```

**Comment by Ph0tonic** [comment]
Great thank you for your feedback, I appreciate it 👍. I implemented a default value of -1 which allows to support no indentation at all if the user provides `0`. I think we can keep the tab as the default and then users can configure whatever they prefer. Let's see what other maintainers think of it.

File: cmd/sops/main.go
Comment (felixfontein) [inline cmd/sops/main.go:1075]: I think this should also work:
```suggestion
	if context.IsSet("indent") {
```
(see https://github.com/urfave/cli/blob/v1-maint/context.go#L55)

The main advantage is that this allows to pass `--indent 0` on CLI and make that override a non-zero indentation for JSON. (For YAML, it will obviously not work.)

File: README.rst
Comment (felixfontein) [inline README.rst:1151]:
```suggestion
this default behaviour to use ``spaces`` by either using the additional ``--indent=2`` CLI option or
```

File: README.rst
Comment (felixfontein) [inline README.rst:1165]:
```suggestion
this default behaviour by either using the additional ``--indent=2`` CLI option or
```

File: README.rst
Comment (felixfontein) [inline README.rst:1166]:
```suggestion
by configuring ``.sops.yaml`` with:
```

**Review by felixfontein — COMMENTED** [review]
I only noticed some small typographic details in the README, besides that it looks good to me!

File: README.rst
Comment (felixfontein) [inline README.rst:1156]: Without that newline, all the content are parameters and the directive has no content, thus showing an empty code listing.
```suggestion
.. code:: yaml

```

File: README.rst
Comment (felixfontein) [inline README.rst:1168]:
```suggestion
.. code:: yaml

```

**Review by felixfontein — COMMENTED** [review]
Sorry, noticed some more docs things ;-)

**Comment by oaxiento** [comment]
It'd be great to see this get merged soon. We really need that feature. Thanks @Ph0tonic @felixfontein @devstein for your effort!

**Comment by hiddeco** [comment]
The plan is to get a patch release out first, before this will be available in the `v3.9.0` release. Thanks for your patience :bow:

File: README.rst
Comment (nogweii) [inline README.rst:1150]: Uh, I believe this is supposed to be "sops indents *JSON* files", no?

Comment (Ph0tonic) [reply README.rst]: Nice catch, thank you !

File: README.rst
Comment (felixfontein) [inline README.rst:206]: Can you move unrelated (to the functionality) fixes to another PR (like htis one), and squash some of the commits? Thanks!

**Comment by felixfontein** [comment]
> The plan is to get a patch release out first, before this will be available in the `v3.9.0` release. Thanks for your patience 🙇

@hiddeco do you already know whether you want to do another patch release before starting to merge new features for 3.9.0?

File: README.rst
Comment (Ph0tonic) [reply README.rst]: Done in #1339

**Comment by hiddeco** [comment]
I think it should be fine to start merging new features, as I was planning on doing `v3.9.0` next.

**Review by hiddeco — COMMENTED** [review]
It would be great if this could be rebased (to get rid of any merge commits), while also smashing some of the e.g. `Update README.rst` commits, etc. together.

For the rebase:

```sh
git remote add upstream https://github.com/getsops/sops
git fetch upstream && git rebase upstream/main
```

For tidying the commits (https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History):

```
git rebase -i upstream/main
```

**Comment by Ph0tonic** [comment]
> It would be great if this could be rebased (to get rid of any merge commits), while also smashing some of the e.g. `Update README.rst` commits, etc. together.
> 
> For the rebase:
> 
> ```shell
> git remote add upstream https://github.com/getsops/sops
> git fetch upstream && git rebase upstream/main
> ```
> 
> For tidying the commits (https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History):
> 
> ```
> git rebase -i upstream/main
> ```

Thanks for helping, here it is :+1:

**Comment by felixfontein** [comment]
Note that commit 07c06c7ac47f68b33b9ef9cbd7f66b6d29d1e953 contains some merge artefacts that are removed again in commit c7b04e1b156699d29393f5043cda1a732fdd7bbe. Would you mind rebasing again to clean up both commits so that the merge artefacts are not part of the commits added to the `main` branch?

Besides that, I think this PR is ready for merging. @hiddeco @devstein do you have further comments?

**Comment by Ph0tonic** [comment]
> Note that commit [07c06c7](https://github.com/getsops/sops/commit/07c06c7ac47f68b33b9ef9cbd7f66b6d29d1e953) contains some merge artefacts that are removed again in commit [c7b04e1](https://github.com/getsops/sops/commit/c7b04e1b156699d29393f5043cda1a732fdd7bbe). Would you mind rebasing again to clean up both commits so that the merge artefacts are not part of the commits added to the `main` branch?
> 
> Besides that, I think this PR is ready for merging. @hiddeco @devstein do you have further comments?

Fixed ! 👍

File: stores/json/store.go
Comment (felixfontein) [inline stores/json/store.go:255]:
```suggestion
	} else if store.config.Indent < -1 {
	    ... tell user that negative values != -1 are not OK ...
	}
```

File: config/config.go
Comment (felixfontein) [inline config/config.go:74]: Shouldn't this also have an `indent` setting?

**Review by felixfontein — COMMENTED** [review]
While looking at this again I have some more comments (or maybe I forgot some context from before).

File: config/config.go
Comment (Ph0tonic) [reply config/config.go:74]: Yes, you are right, I will add this parameter

Comment (Ph0tonic) [reply config/config.go]: Here it is, I implemented it this way so we can have a different configuration for `json` and `json_binary`. Let me know what you think.

File: cmd/sops/main.go
Comment (felixfontein) [inline cmd/sops/main.go:1093]:
```suggestion
		storesConf.JSON.Indent = indent
		storesConf.JSONBinary.Indent = indent
```

**Review by felixfontein — COMMENTED** [review]
Besides that, looks good to me!

File: cmd/sops/main.go
Comment (Ph0tonic) [reply cmd/sops/main.go:1093]: Nice catch, thanks.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this during the next few days.

**Review by hiddeco — APPROVED** [review]

**Comment by felixfontein** [comment]
@jamesgoodhouse @Ph0tonic thanks a lot for working on this and getting it over the finish line! :)

**Comment by consideRatio** [comment]
Wieee happy to learn that the json files are now being valid YAML files by no longer including tabs! Nice work on this @Ph0tonic @jamesgoodhouse and reviewers! :tada:

**Pull request merged**
